packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,123 @@
+import * as cdk from '@aws-cdk/core';
+import { TagOption } from '../common';
+import { TagUpdatesOptions } from '../constraints';
+import { IPortfolio, Portfolio } from '../portfolio';
+import { IProduct, Product } from '../product';
+import { CfnPortfolioProductAssociation, CfnResourceUpdateConstraint, CfnTagOption, CfnTagOptionAssociation } from '../servicecatalog.generated';
+import { hashValues } from './util';
+import { InputValidator } from './validation';
+
+export class AssociationManager {
+  public static associateProductWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct) {
+    const associationKey = hashValues(portfolio.node.addr, product.node.addr, scope.stack.node.addr);
+    if (!this.associationMap.has(associationKey)) {
+      const association = new CfnPortfolioProductAssociation(scope, `PortfolioProductAssociation${associationKey}`, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+      this.associationMap.set(associationKey, new Map());
+      this.associationMap.get(associationKey).set(Constraints.ASSOCIATION, association);
+    }
+  }
+
+  public static addResourceUpdateConstraint(scope: cdk.Resource, product: IProduct, options: TagUpdatesOptions) {
+    const [pair, associationKey] = this.associationPrecheck(scope, product);
+    InputValidator.validateLength(this.generateAssocationString(scope, pair), 'description', 0, 2000, options.description);
+    if (!this.associationMap.get(associationKey).get(Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.acceptLanguage,
+        description: options.description,
+        portfolioId: pair.portfolio.portfolioId,
+        productId: pair.product.productId,
+        tagUpdateOnProvisionedProduct: options.tagUpdateOnProvisionedProductAllowed || options.tagUpdateOnProvisionedProductAllowed === undefined
+          ? Allowed.ALLOWED : Allowed.NOT_ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey).get(Constraints.ASSOCIATION));
+      this.associationMap.get(associationKey).set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.generateAssocationString(scope, pair)}`);
+    }
+  }
+
+  public static associateTagOption(scope: cdk.Resource, resourceId: string, tagOption: TagOption) {
+    Object.keys(tagOption).forEach(key => {
+      InputValidator.validateLength(resourceId, 'TagOption key', 1, 128, key);
+      tagOption[key].forEach(value => {
+        InputValidator.validateLength(resourceId, 'TagOption value', 1, 256, value.value);
+        const tagOptionKey = hashValues(key, value.value);
+        if (!this.tagOptionMap.has(tagOptionKey)) {
```

Comment:
When are there duplicates already in this map? Doesn't this mean the expression below:


```ts
        new CfnTagOptionAssociation(scope, `TagOptionAssociation${hashValues(scope.node.addr, tagOptionKey)}`, {
```

Would fail, because there's a duplicate identifier in `scope`?

Reply:
correct, cdk handles the error basically in the same way we would, telling user there is already a resource with that name and provides the name, wasn't sure if I should wrap that and basically throw same error again.

Reply:
You have an `if` here already... seems that throwing an error with a nicer message than "duplicate construct ID" is what you want here...

Reply:
Duplicates would occur if you attempt to add the same key-value pair multiple times.

Reply:
Looking back on this it would align more with strategy to make this idempotent.  If an end-user is trying to associate  tagOption, we should handle if they do it twice not break.

Reply:
Agreed.